services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java

```diff
@@ -473,30 +473,33 @@ public Response createTokenResponse(UserModel user, UserSessionModel userSession
             responseBuilder.generateIDToken().generateAccessTokenHash();
         }
 
+        AccessTokenResponse res = new AccessTokenResponse();
         if (clientPolicyContextGenerator != null) {
             try {
-                session.clientPolicy().triggerOnEvent(clientPolicyContextGenerator.apply(responseBuilder));
+                session.clientPolicy().triggerOnEvent(clientPolicyContextGenerator.apply(responseBuilder, res));
```

Comment:
What is the reason for passing a reference to `AccessTokenResponse` and not just using the response builder as it works today?

Reply:
The current client policies can already customize a token response as customizing a response builder.
The PR extends this customizability that can customize the response as the way that cannot be realized by only a response builder.

Reply:
I suppose that main reason for this is, that you want some clientPolicyExecutor to be able to build the response by itself. I guess that is also the reason for the above `if` check `if (res.getToken() == null) {` , so the response is not built another time.

One alternative approach to this might be to slightly refactor `TokenManager.AccessTokenResponseBuilder`, so that it will allow duplicated call to `build()` . I am thinking about something like:
- In `TokenManager.AccessTokenResponseBuilder`, introducing variable like:
```
private AccessTokenResponse response;
```

- Then refactor method `TokenManager.AccessTokenResponseBuilder.build()` to be like:
```
public AccessTokenResponse build() {
   if (this.response != null) return this.response;
   ... Do the same code as of today with the exception of last line ...
   this.response = res;
   return this.response; 
```
In this case, there won't be a need to pass empty AccessTokenResponse and also it would be possible to avoid `if` condition below.

WDYT?